### DISCUSSION TITLE
Can Mojo 🔥 be used to develop microcontrollers such as Arduino?

### DISCUSSION THREAD

**Discussion by Eprahim-taha** [modular/modular#1461] [Q&A]
I wonder if the Mojo programming language 🔥 is capable of programming microcontrollers like the Arduino?

**Comment by rarebreed** [comment]
Eventually, it could depending on how coupled mojo is with its standard library.  

Rust has a `#![no_std]` capability so that binaries will not be built with the regular `std`.  Many of rust's data types are built on top of primitives or other types defined by the standard library.  In turn, the standard library has hooks into libc and other system interface (ie POSIX) primnitives, so that it can get access to system calls.  That in turn gives you access to OS resources...like memory including mem-mapped IO.  If you're programming on bare metal machines, you probably won't have these system interfaces (though you may have a more minimal C lib interface).

Mojo would probably need to have a very basic std-less core that the regular standard library rests on top of.  I haven't programmed on arduino's, but I know the concept with other baremetal devices.  I'm guessing arduino already comes with a library to do things like allocate memory and do memory-mapped IO (so you can access GPIO pins).  Mojo should be able to access the library with some FFI to call C functions (although I'm not sure what kind of ABI arduino has)
